src/main/thrift/parquet.thrift

```diff
@@ -1165,6 +1317,62 @@ struct FileMetaData {
   9: optional binary footer_signing_key_metadata
 }
 
+/** Metadata for a column in this file. */
+struct FileColumnMetadataV3 {
+  /** All column chunks in this file (one per row group) **/
+  1: required list<ColumnChunkV3> columns
```

Comment:
See comment below, I think this  should like be an offset(S) to a data page that individually encodes the information in columns.

Reply:
I agree that a Parquet designed from scratch would perhaps store most of its metadata using regular data columns and encodings (just like databases do). But evolving from the existing Parquet format is a slightly different problem. The more the new format structures deviate from the legacy ones, the more work we will have to implement it and the more pains we will suffer.

Reply:
So I am not advocating full shredding (e.g. a sub parquet schema for each one of these elements), I agree that would be overdone and most likely more of a perfomance hindrance.

I do think moving away from using thrifts lists as much as possible will help significantly in overhead reduction.  The issue here is that you cannot selectively decode part of lists in an easy manner.  I think a design goal here is we should be trying to make deserialization cost O(columns scanned) vs O(number of column in schema)

Reply:
@emkornfield Thanks for that comment emkornfield! This is the bottleneck we are struggling with, as per previous PR, is getting fast access random rowgroup reads. Having the columns (and rowgroups below) stored in a fashion that allows random access would be a huge improvement here!

Reply:
A few other throughts here:
1.  Once we start having different messages here in a lot of languages the extra development effort to be able to read a random access page and extra a byte-array probably isn't too high compared to having completely new types to parse and translate (i.e. there is going to be a fair bit of boiler plate).
2. By treating list as a black-box of bytes, it leaves open the possibility of changing encodings for the elements if we have data showing that is a large improvement.
3. Instead of being an offset, I suppose this could just be modeled in the message as [bytes], the main downside to this is thrift bindings that don't allow zero copy.

Reply:
> Instead of being an offset, I suppose this could just be modeled in the message as [bytes]

Sure, but would that change exactly? You'll have to decode it anyway.

Reply:
> > Instead of being an offset, I suppose this could just be modeled in the message as [bytes]
> 
> Sure, but what would that change exactly? You'll have to decode it anyway.

If the question is specifically for bytes in thrift vs a page stored someplace else, I think the main trade-off is how easy it is to achieve zero copy on the underlying bytes that need to be decoded vs the complexity of handling the offset.  Decoding of just one a few large byte arrays does not have to be expensive, but in practice at least in C++ it is more expensive then it probably should be.  It seems java might actually have the ability to do zero copy here.

If the question is why a data page (which still needs decoding), I was thinking if we introduced a new encoding which effectively uses the arrow byte array layout `([cumulative_offsets_into_byte_data], [byte_data])`, then decoding is really the of parsing the page header, and the page would provide random access to elements that could be decoding individually.  For end-users more concerned about overall space, then the page abstraction still allows for more complex encodings/compression to reduce footer size.

Reply:
@emkornfield One additional point in favor of a set of bytes may be random access for subsets of the metadata. One big problem with the current thrift structure is that we don't really know the (variable) size of a column chunk and have to thrift decode everything. A byte array with either a fixed size or a length attribute would be one way to get random access to column chunks.

Reply:
I think that if instead of:
 `required list<ColumnChunkV3> columns`
we had another indirection
`required list<i64> column_chunk_offsets // n+1 offsets to ColumnChunkV3`
it would solve the problem of random row group access.
With another level of indirection the data would be more fragmented in the footer - but with the advantage of much cheaper initial decoding.

Reply:
> It could also be a required list<binary> encoded_chunks  // Thrift-encoded ColumnChunkV3 objects. 

I think this leaves us at the mercy of thrift parsers with respect to additional copies (which might not be huge but seems to be something we can avoid if we move the list to a "page")

Reply:
> > I think that if instead of:
> > `required list<ColumnChunkV3> columns`
> > we had another indirection
> > `required list<i64> column_chunk_offsets // n+1 offsets to ColumnChunkV3`
> > it would solve the problem of random row group access.
> 
> It could also be a `required list<binary> encoded_chunks // Thrift-encoded ColumnChunkV3 objects`. That way, you don't have to issue one additional IO for each column chunk.

FYI: I've run a quick experiment to see the difference between thrift-decoding of 1 million elements of different types. 
According to my results, binary is about 10-15x faster than the struct and the offset is 100x faster than the struct.

My test thrift structure was:
```
struct MyTestStructure {
  1: optional list<ColumnChunk> columns_struct
  2: optional list<binary> columns_binary
  3: optional list<i64> column_offsets
}
```

The results are:
```
 Deserialize structs: 41000006 bytes, 409ms
 Deserialize binary: 42000006 bytes, 25ms
 Deserialize offsets: 2991750 bytes, 4ms
 ```

Reply:
Thanks for the numbers @marcin-krystianc (which Thrift implementation did you test with?). 
That's an overhead of about 25ns per binary vs. 4ns per offset-pair. Given that the inlined binary spares an additional indirect IO, this seems like a reasonable tradeoff to make.

Reply:
> Thanks for the numbers @marcin-krystianc (which Thrift implementation did you test with?). 

I used thrift from apt repository (`thrift-compiler` package, `Thrift version 0.16.0`)
The compilation command was: `thrift -r --gen cpp:no_default_operators,no_ostream_operators`

Reply:
> I think this leaves us at the mercy of thrift parsers with respect to additional copies

This is rather strong wording :-) Memory copies are not the bane of performance, assuming they remain small-ish.

> which might not be huge but seems to be something we can avoid if we move the list to a "page"

At the cost of at least one additional indirect IO for each column chunk, which might not always be desirable?

---

I'm trying to find a good tradeoff here. I think we have three reasonable options:
1. Use file offsets for both row groups and column chunks.
2. Use file offsets for row groups, and embedded Thrift-encoded binary for column chunks.
3. Use embedded Thrift-encoded binary for row groups, and file offsets for column chunks.

File metadata load speed should be fast for options 1 and 2, less so for option 3.

Accessing the metadata for M columns in N row groups requires _O(M * N)_ IOs in option 1; _O(N)_ IOs in option 2; _O(M)_ IOs in option 3. Since N is typically much smaller than M (a row group is typically very large, you don't need to access many at once), it seems that option 2 is better than option 3 here, though both are much better than option 1.

_However_, some readers might be able to coalesce IOs together if they are contiguous enough. In this case, the difference between M and N would matter much less and options 2 and 3 could end up issuing a similar number of IOs to access column chunk metadata.

I would therefore intuitively tend towards option 2, but that's not backed by numbers.

Did I get something wrong in these speculations?

Reply:
If the "desired algorithm" is really to preload all metadata at once, then I'm not sure why we would build any indirections into the format. I'm also not sure what a "metadata page" is, it seems like you're running with your own idea instead of discussing what's being proposed here :-)

> undo complexity I guess is in the eye of the beholder

More importantly, the offset/length scheme implies a back-to-back dependency between IO operations.

Reply:
Yes, I think I am expressing an alternative idea.  Let try to draft a proposal in a little more depth instead of sprinkling possible hard to understand comments on this PR

Reply:
https://github.com/apache/parquet-format/pull/250